openedx/core/djangoapps/safe_sessions/tests/test_middleware.py

```diff
@@ -615,3 +622,725 @@ def test_user_change_with_no_ids(self):
+
+    @override_settings(ENFORCE_SESSION_EMAIL_MATCH=False)
+    @patch('openedx.core.djangoapps.safe_sessions.middleware._mark_cookie_for_deletion')
+    @patch("openedx.core.djangoapps.safe_sessions.middleware.set_custom_attribute")
+    def test_process_request_emails_mismatch_with_toggle_disabled(
+        self, mock_set_custom_attribute, mock_mark_cookie_for_deletion
+    ):
+        """
+        Calls EmailChangeMiddleware.process_request when user is authenticated,
+        ENFORCE_SESSION_EMAIL_MATCH is disabled and user session and request
+        email mismatch. (Email was changed in some other browser)
+        Verifies that session and cookies are not affected.
+        """
+        # Log in the user
+        self.client.login(email=self.user.email, password=self.PASSWORD)
+        self.request.session = self.client.session
+        self.client.response.set_cookie(settings.SESSION_COOKIE_NAME, 'authenticated')  # Add some logged-in cookie
+
+        # Registering email change (store user's email in session for later comparison by
+        # process_request function of middleware)
+        EmailChangeMiddleware.register_email_change(request=self.request, email=self.user.email)
+
+        # Ensure email is set in the session
+        self.assertEqual(self.request.session.get('email'), self.user.email)
+        # Ensure session cookie exist
+        self.assertEqual(len(self.client.response.cookies), 1)
+
+        # simulating email changed in some other browser
+        self.user.email = 'new_email@test.com'
+        self.user.save()
+
+        # Call process_request
+        EmailChangeMiddleware(get_response=lambda request: None).process_request(self.request)
+
+        # Verify that session_email_mismatch and is_enforce_session_email_match_enabled
+        # custom attributes are set
+        mock_set_custom_attribute.assert_has_calls([call('session_email_mismatch', True)])
+        mock_set_custom_attribute.assert_has_calls([call('is_enforce_session_email_match_enabled', False)])
+
+        # Assert that the session and cookies are not affected
+        self.assertNotEqual(self.request.session.get('email'), self.user.email)
+        self.assertEqual(len(self.client.response.cookies), 1)
+        self.assertEqual(self.client.response.cookies[settings.SESSION_COOKIE_NAME].value, 'authenticated')
+
+        # Assert that _mark_cookie_for_deletion not called
+        mock_mark_cookie_for_deletion.assert_not_called()
+
+    @override_settings(ENFORCE_SESSION_EMAIL_MATCH=True)
+    @patch('openedx.core.djangoapps.safe_sessions.middleware._mark_cookie_for_deletion')
+    def test_process_request_no_email_change_history_with_toggle_enabled(
+        self, mock_mark_cookie_for_deletion
+    ):
+        """
+        Calls EmailChangeMiddleware.process_request when there is no previous
+        history of an email change and ENFORCE_SESSION_EMAIL_MATCH is enabled
+        Verifies that existing sessions are not affected.
```

Comment:
Its indeed a very suitable suggestion and I have implemented that. Could you please have a final look on the PR. Thanks

Reply:
It is a very suitable suggestion and yes its an alternative to what we are doing in registration/login.

But there is an issue with that, by this change we get 2 tests failing. ([test_too_many_courses](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_api/tests/test_views.py#L405) and [test_get_unknown_course](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_home_api/course_metadata/tests/test_views.py#L88).
I was not able to fully understand the reasoning for them but `test_too_many_courses` gives an error `executed 53 queries instead of 50` and my debugging shows that the additional 3 queries executed were of session updation. So it seems to be that having `request.session['email'] = request.user.email` in the middleware is adding extra queries and failing those tests? [I have implemented this solution in my PR and these tests are failing]

Could you please guide on how to resolve this? I would appreciate your opinion/suggestion on this.

Reply:
1. I think it is ok to simply adjust 50 to 53 for the first test if we know why it is changing. The query assertions are there to ensure that the number of queries doesn't grow without explanation, or grow very large.
2. Do you understand why 404 => 400? That seems a little more scary to me without understanding why that is changing.

Reply:
If the new code works for registration, and this might just be a strange case in tests, might it make sense to wrap, swallow, and log/set_custom_attribute for any exception happening for the save, and then continue on rather than raising the exception?

Reply:
> might it make sense to wrap, swallow, and log/set_custom_attribute for any exception happening for the save, and then continue on rather than raising the exception?

Unfortunately, I don't think we can wrap the `request.session.save()` for exception since its inside the `SessionMiddleware.process_response` which is inherited by `SafeSessionMiddleware`. 

Some further debugging shows that its probably due to the `raise Http404()`.  `request.session.save()` fails if view calls `raise Http404()` and works fine if the view returns normal `HttpResponse`. 
I also tried reproducing the test's flow locally but it was working as expected (returning 404 instead of 400). I am still unclear on the reason why `request.session.save()` getting failed in the test but i am trying to find the reason.

Reply:
Further findings

The Django test cases, specifically the test case in question, are wrapped inside `transaction.atomic()`. When an error occurs within the transaction (such as raising `Http404()` in our case), it leads to the failure of subsequent database operations (e.g., `request.session.save()`). This is confirmed by the following error in stack trace of test failure.

```
raise TransactionManagementError(
django.db.transaction.TransactionManagementError: An error occurred in the current transaction. You can't execute queries until the end of the 'atomic' block.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/sessions/middleware.py", line 61, in process_response
    request.session.save()
  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/django/contrib/sessions/backends/db.py", line 94, in save
    raise UpdateError
django.contrib.sessions.backends.base.UpdateError
```

Now this answers the questions I guess

**Why was it happening for test only not in real execution?**
Because only tests are wrapped inside `transaction.atomic()` to make sure that db transactions can be rolled back at the end of tests.

**Why did it occur for `Http404` but not for normal `HttpResponseForbidden`?**
`Http404()` is an exception, causing `transaction.atomic()` to fail, while `HttpResponseForbidden` is a regular response with a 404 status code.

**Why were there database transactions in tests?**
Django automatically switches to a database session store for tests, ensuring an isolated and consistent testing environment.

**Why didn't it occur for successful responses?**
Success responses were instances of `HttpResponse` and not exceptions thus not causing the transaction failure.

Reply:
Based on this, i think we would need to move the `request.session['email'] = request.user.email` back to login/register views. What do you think?

Reply:
As discussed in our meeting, we think we have the right solution and you will experiment with finding the best way to adjust tests. We'll see what you find.

Reply:
My findings:

[test_get_unknown_course](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_home_api/course_metadata/tests/test_views.py#L88) was the only test having the `TransactionManagement` error. Other tests (around 4 to 5 in count) only needed query count adjustments. 

Currently, I have used the approach of having the settings flag for that test (as we discussed). But I have found out that `test_get_unknown_course` is the only test getting affected by the `TransactionManagement` error. 
I was wondering if we could simply add `self.client.logout()` in that test and make that work? That test is just checking/testing that the view returns `404` for an invalid course ID. I have verified that the view returns `404` for invalid course IDs for both authenticated and unauthenticated users. 

This way we wouldn't have to define a settings flag just for one test (that too isn't majority concerned with user authentication and all the session saving)

Example View URL (with invalid course id):  https://courses.edx.org/api/course_home/course_metadata/course-v1:HarvardX+CS50+X --- you can try hitting this without having a logged-in user.

What are your thoughts on this?

Reply:
I think that sounds fine. You could add a brief comment about this, or not if you think it would just be confusing.

Reply:
Hi @robrap,

I have updated the PR. Could you please have a final look?

Reply:
Marking resolved. I'll add updated comments as-needed.